instrumentation/kafka-clients/src/main/java/brave/kafka/clients/TracingConsumer.java

```diff
@@ -0,0 +1,228 @@
+package brave.kafka.clients;
+
+import brave.Span;
+import brave.Tracing;
+import brave.propagation.TraceContext;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Pattern;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRebalanceListener;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.clients.consumer.OffsetAndTimestamp;
+import org.apache.kafka.clients.consumer.OffsetCommitCallback;
+import org.apache.kafka.common.Metric;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+
+/**
+ * Kafka Consumer decorator.
+ * Read records headers to create and complete a child of the incoming producers span if possible.
+ */
+class TracingConsumer<K, V> implements Consumer<K, V> {
+
+  private final KafkaPropagation.ConsumerInjector injector =
+      new KafkaPropagation.ConsumerInjector();
+  private final KafkaPropagation.ConsumerExtractor extractor =
+      new KafkaPropagation.ConsumerExtractor();
+
+  private final Tracing tracing;
+  private final Consumer<K, V> wrappedConsumer;
+
+  TracingConsumer(Tracing tracing, Consumer<K, V> consumer) {
+    this.wrappedConsumer = consumer;
+    this.tracing = tracing;
+  }
+
+  /**
+   * For each poll, handle span creation.
+   */
+  @Override
+  public ConsumerRecords<K, V> poll(long timeout) {
+    ConsumerRecords<K, V> records = wrappedConsumer.poll(timeout);
+    for (ConsumerRecord<K, V> record : records) {
+      handleConsumed(record);
+    }
+    return records;
+  }
+
+  private void handleConsumed(ConsumerRecord record) {
+    Span span = startAndFinishConsumerSpan(record);
+    // remove propagation headers from the record
+    tracing.propagation().keys().forEach(key -> record.headers().remove(key));
+
+    // Inject new propagation headers
+    tracing.propagation().injector(injector)
```

Comment:
save some overhead by declaring injector as a field and reusing it. `injector = tracing.propagation().injector(injector)`